### PR TITLE
hoc2021: Add new scripts to list of HOC script constants

### PR DESCRIPTION
**Pull request by breville**
This adds the "I finished!" link to the header above these scripts, and should add the scripts to the course assignment dropdown once they are each taken out of beta.

### CODE REVIEW

**Review by maddiedierker — COMMENTED** [review]
this change looks good, but it looks like reordering this list caused a test to fail. it may have been a transient issue, but i wonder if this list is somehow order dependent?

**Review by KylieModen — APPROVED** [review]
LGTM, but +1 Maddie's comment - want to make sure the reorder doesn't break anything

**Comment by breville** [comment]
I've updated the test now that @KylieModen has confirmed the ordering.  (Not a huge fan of tests that re-express dynamic design decisions, but not a big deal.)

File: lib/cdo/script_constants.rb [hunk 1]
```diff
@@ -74,10 +74,17 @@ module ScriptConstants
...
     hoc: [
       # Note that now multiple scripts can be an 'hour of code' script.
-      # If adding a script here,
-      # you must also update the cdo-tutorials gsheet so the end of script API works
+      # If adding a script here, you must also update the cdo-tutorials gsheet
+      # so the end of script API works; specifically, there needs to be a row
+      # with code_s matching the script name (in quotes) in this list.
 
       nil,
+      POEM_ART = 'poem-art'.freeze, # 2021 hour of code
```
Comment (bethanyaconnor) [inline lib/cdo/script_constants.rb]: These names are planning on being updated to append `-2021` right? Any reason to make this change before those names are updated?

Comment (breville) [reply lib/cdo/script_constants.rb]: I just learned about that and am happy to hold off until they are ready.

Comment (dmcavoy) [reply lib/cdo/script_constants.rb]: Hey @breville - It looks like I will be making those copies After 12pm PT on 11/11. I will keep you updated as I make the updates. I could also just take this off your hands if you want and include it as part of that work?

Comment (breville) [reply lib/cdo/script_constants.rb]: @dmcavoy If you'd like to take over this PR, that would be great.  We don't want these tutorials appearing in the course assignment dropdown until next week; I'm still curious when this PR can be merged to ensure that.

File: lib/cdo/script_constants.rb [hunk 2]
```diff
@@ -268,17 +274,11 @@ module ScriptConstants
     DEEPDIVE_DEBUGGING = 'deepdive-debugging'.freeze,
     FREQUENCY_ANALYSIS = 'frequency_analysis'.freeze,
     GAMELAB = 'gamelab'.freeze,
-    HELLO_WORLD_FOOD = 'hello-world-food'.freeze,
```
Comment (dmcavoy) [inline lib/cdo/script_constants.rb]: I'm not sure if we want to remove these from the i18n list?

Comment (bethanyaconnor) [reply lib/cdo/script_constants.rb]: This is fine because all [hoc scripts are translated](https://github.com/code-dot-org/code-dot-org/blob/659876989008c7e2548933b0669fec5c84d877d4/lib/cdo/script_constants.rb#L459)

Comment (dmcavoy) [reply lib/cdo/script_constants.rb]: Great! Thank you Bethany.

Comment (bethanyaconnor) [reply lib/cdo/script_constants.rb]: ooo but if we keep these ones around for translation we probably do want to keep them them (sorry for the turn around, I'm having trouble keeping all of this straight). But the *-2021 scripts don't need to be added here

Comment (dmcavoy) [reply lib/cdo/script_constants.rb]: Oh ya good point! Need these and then to add the -2021 ones above. I will update this PR once everything is copied

File: lib/cdo/script_constants.rb [hunk 3]
```diff
@@ -74,10 +74,16 @@ module ScriptConstants
...
+      # If adding a script here, you must also update the cdo-tutorials gsheet
+      # so the end of script API works; specifically, there needs to be a row
+      # with code_s matching the script name (in quotes) in this list.
 
       nil,
+      POEM_ART_2021_NAME = 'poem-art-2021'.freeze, # 2021 hour of code
+      HELLO_WORLD_FOOD_2021_NAME = 'hello-world-food-2021'.freeze, # 2021 hour of code
+      HELLO_WORLD_ANIMALS_2021_NAME = 'hello-world-animals-2021'.freeze, # 2021 hour of code
+      HELLO_WORLD_EMOJI_2021_NAME = 'hello-world-emoji-2021'.freeze, # 2021 hour of code
+      HELLO_WORLD_RETRO_2021_NAME = 'hello-world-retro-2021'.freeze, # 2021 hour of code
```
Comment (dmcavoy) [inline lib/cdo/script_constants.rb:93]: @breville @KylieModen This is only 5 of the 10 scripts I was askd to copy over. Is that correct? Do we want the other new scripts in the list too?

Comment (dmcavoy) [reply lib/cdo/script_constants.rb:93]: Script List https://docs.google.com/spreadsheets/d/1mFFYScB9wmGL59rqXF8uEcQl-IYhB9duH8nBc4P4A0g/edit#gid=0

Comment (KylieModen) [reply lib/cdo/script_constants.rb:93]: The other scripts are CSC (not HoC) so they don’t need this treatment

Comment (dmcavoy) [reply lib/cdo/script_constants.rb:93]: If we want them to be assignable they will need to be added to script constants somewhere. Probably not in the HOC list but as their own course like we do for CSD/CSP etc

File: lib/cdo/script_constants.rb [hunk 4]
```diff
@@ -74,10 +74,16 @@ module ScriptConstants
...
       # Note that now multiple scripts can be an 'hour of code' script.
-      # If adding a script here,
-      # you must also update the cdo-tutorials gsheet so the end of script API works
+      # If adding a script here, you must also update the cdo-tutorials gsheet
+      # so the end of script API works; specifically, there needs to be a row
+      # with code_s matching the script name (in quotes) in this list.
 
       nil,
+      POEM_ART_2021_NAME = 'poem-art-2021'.freeze, # 2021 hour of code
+      HELLO_WORLD_FOOD_2021_NAME = 'hello-world-food-2021'.freeze, # 2021 hour of code
```
Comment (dmcavoy) [inline lib/cdo/script_constants.rb:90]: @KylieModen @mikeharv All of the Hello World scripts have the same name "Hello World" so there is nothing differentiating them in the course assignment dropdown

Screen Shot 2021-11-11 at 10 17 39 PM

File: lib/cdo/script_constants.rb (hunk 3 above)
Comment (dmcavoy) [reply lib/cdo/script_constants.rb:93]: Actually they may just show up in the "Other" list it looks like. Is that what we want? Or do we want a CSC section?

File: lib/cdo/script_constants.rb [hunk 5]
```diff
@@ -274,11 +274,17 @@ module ScriptConstants
...
+    HELLO_WORLD_ANIMALS = 'hello-world-animals'.freeze,
+    HELLO_WORLD_EMOJI = 'hello-world-emoji'.freeze,
+    HELLO_WORLD_RETRO = 'hello-world-retro'.freeze,
     K1HOC_2017 = 'k1hoc2017'.freeze,
     NETSIM = 'netsim'.freeze,
     ODOMETER = 'odometer'.freeze,
     OUTBREAK = 'outbreak'.freeze,
     PIXELATION = 'pixelation'.freeze,
+    POEM_ART = 'poem-art'.freeze,
+    POETRY_HOC3 = 'poetry-hoc3'.freeze,
```
Comment (breville) [inline lib/cdo/script_constants.rb]: Do we still use `poetry-hoc3`?  It sounds a bit like a temporary name, but I'm not sure its history.

Comment (dmcavoy) [reply lib/cdo/script_constants.rb]: I don't know the answers but I would be worried that we might lose translations by deleting it. We are planning to do a pass to deleted the old scripts from HOC so I think this can be cleaned up when we do

Comment (bethanyaconnor) [reply lib/cdo/script_constants.rb]: This was added to get an early start on translations. I believe it was superseded by poem-art, which will in turn be superseded by poem-art-2021. I'd want to check with the foundations team that translations won't be lost before we remove it. It _should_ be safe as poem-art should have been synced for a couple of weeks

**Comment by dmcavoy** [comment]
Please hold on review actually. I need to make a category for CSC and put the scripts in it

File: lib/cdo/script_constants.rb (hunk 4 above)
Comment (dmcavoy) [reply lib/cdo/script_constants.rb:90]: Resolved by updating the names in levelbuilder

File: lib/cdo/script_constants.rb (hunk 3 above)
Comment (dmcavoy) [reply lib/cdo/script_constants.rb:93]: Resolved by adding a new CSC section

**Comment by dmcavoy** [comment]
Update of what it will generally look like. Note this is on my local machine and 3 of the scripts have not made it to staging and the Hello World name changes also have not made it to staging

Screen Shot 2021-11-12 at 4 30 45 PM

**Comment by dmcavoy** [comment]
I'm going to pull the category name change out into another PR as I think we will want that to go out today before the Sunday Scoop but we don't want other things to be included in the deploy until Monday

File: dashboard/config/locales/scripts.en.yml [hunk 6]
```diff
@@ -25,6 +25,7 @@ en:
         cspexams_category_name: CS Principles Practice Test
         csp17_category_name: CS Principles ('17-'18)
         aiml_2021_category_name: AI / ML
+        csc_2021_category_name: CS Connnections ('21-'22)
```
Comment (maddiedierker) [inline dashboard/config/locales/scripts.en.yml:28]: you'll probably have to undo this change to prevent a merge conflict with https://github.com/code-dot-org/code-dot-org/pull/43580

File: lib/test/test_script_constants.rb [hunk 7]
```diff
@@ -62,18 +62,19 @@ def test_category_index
...
   def test_assignable_info
-    assert_equal 1, ScriptConstants.assignable_info({name: 'dance-2019'})[:position]
-    assert_equal 2, ScriptConstants.assignable_info({name: 'dance-extras-2019'})[:position]
-    assert_equal 3, ScriptConstants.assignable_info({name: 'oceans'})[:position]
-    assert_equal 4, ScriptConstants.assignable_info({name: 'aquatic'})[:position]
-    assert_equal 5, ScriptConstants.assignable_info({name: 'hero'})[:position]
-    assert_equal 6, ScriptConstants.assignable_info({name: 'mc'})[:position]
-    assert_equal 7, ScriptConstants.assignable_info({name: 'minecraft'})[:position]
+    dance_index = 6
+    assert_equal dance_index, ScriptConstants.assignable_info({name: 'dance-2019'})[:position]
```
Comment (maddiedierker) [inline lib/test/test_script_constants.rb:71]: [not blocking for this PR] i know this test has been around for a while, but i'm not sure how useful it is. it does prevent us from reordering this list (although it only tests the positions of scripts within `CATEGORIES[:hoc]`), but it seems like a warning comment is sufficient, which we already have:

https://github.com/code-dot-org/code-dot-org/blob/8727f5e9117d37ca69fd4fab6159103eb7804fde/lib/cdo/script_constants.rb#L17-L20

and it seems to only fail like it does in this case where we're intentionally changing the positions to add new scripts. 

this test was added in https://github.com/code-dot-org/code-dot-org/pull/19407 to fix a real regression, but i wonder if the case we ran into is still relevant or if we've made improvements elsewhere to avoid that situation from happening again.

just some thoughts 🤷🏻‍♀️

**Review by maddiedierker — APPROVED** [review]

**Review by bethanyaconnor — APPROVED** [review]

File: dashboard/config/locales/scripts.en.yml (hunk 6 above)
Comment (maddiedierker) [reply dashboard/config/locales/scripts.en.yml:28]: i think i was wrong about this -- that PR has been merged and there don't appear to be any conflicts. i suppose because it's the exact same change, github worked it out automatically
